From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lyt99 <lyt1999131@163.com>
Date: Wed, 9 Jun 2021 16:29:32 +0800
Subject: [PATCH] overwrite endpoint when conflicting

Signed-off-by: l1b0k <libokang.dev@gmail.com>
---
 daemon/cmd/endpoint.go | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/daemon/cmd/endpoint.go b/daemon/cmd/endpoint.go
index 9700af0b3a..c7c4ad1637 100644
--- a/daemon/cmd/endpoint.go
+++ b/daemon/cmd/endpoint.go
@@ -357,7 +357,9 @@ func (d *Daemon) createEndpoint(ctx context.Context, owner regeneration.Owner, e
 		if err != nil {
 			return invalidDataError(ep, err)
 		} else if oldEp != nil {
-			return invalidDataError(ep, fmt.Errorf("IP %s is already in use", id))
+			oldEp.Logger("api").Warning("endpoint conflicted, deleting")
+			d.deleteEndpoint(oldEp)
+			// return invalidDataError(ep, fmt.Errorf("IP %s is already in use", id))
 		}
 	}
 
-- 
2.39.1

