From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: l1b0k <libokang.lbk@alibaba-inc.com>
Date: Thu, 18 Dec 2025 11:00:11 +0800
Subject: endpoint: fix sip config

Signed-off-by: l1b0k <libokang.lbk@alibaba-inc.com>
---
 pkg/endpoint/endpoint.go | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/pkg/endpoint/endpoint.go b/pkg/endpoint/endpoint.go
index 291777e2b8..db9c78d8eb 100644
--- a/pkg/endpoint/endpoint.go
+++ b/pkg/endpoint/endpoint.go
@@ -2671,6 +2671,10 @@ func (e *Endpoint) SetDefaultConfiguration() {
 		return
 	}
 	e.setDefaultPolicyConfig()
+
+	if e.DatapathConfiguration.DisableSipVerification {
+		e.Options.SetValidated(option.SourceIPVerification, option.OptionDisabled)
+	}
 }
 
 func (e *Endpoint) setDefaultPolicyConfig() {
-- 
2.50.1 (Apple Git-155)

