From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: l1b0k <libokang.dev@gmail.com>
Date: Thu, 10 Jun 2021 19:20:58 +0800
Subject: [PATCH] adapt 1.10 for terway

Signed-off-by: l1b0k <libokang.dev@gmail.com>
---
 pkg/option/config.go | 20 ++++++++++----------
 1 file changed, 10 insertions(+), 10 deletions(-)

diff --git a/pkg/option/config.go b/pkg/option/config.go
index 7c0eaa8932..f713067030 100644
--- a/pkg/option/config.go
+++ b/pkg/option/config.go
@@ -3193,16 +3193,16 @@ func (c *DaemonConfig) Populate() {
 		}
 	}
 
-	switch c.IPAM {
-	case ipamOption.IPAMKubernetes, ipamOption.IPAMClusterPool, ipamOption.IPAMClusterPoolV2:
-		if c.EnableIPv4 {
-			c.K8sRequireIPv4PodCIDR = true
-		}
-
-		if c.EnableIPv6 {
-			c.K8sRequireIPv6PodCIDR = true
-		}
-	}
+	//switch c.IPAM {
+	//case ipamOption.IPAMKubernetes, ipamOption.IPAMClusterPool, ipamOption.IPAMClusterPoolV2:
+	//	if c.EnableIPv4 {
+	//		c.K8sRequireIPv4PodCIDR = true
+	//	}
+	//
+	//	if c.EnableIPv6 {
+	//		c.K8sRequireIPv6PodCIDR = true
+	//	}
+	//}
 
 	c.KubeProxyReplacementHealthzBindAddr = viper.GetString(KubeProxyReplacementHealthzBindAddr)
 
-- 
2.39.1

