From c2e3084fe5ca96e6df5be2c63acdd82b1ed99197 Mon Sep 17 00:00:00 2001
From: l1b0k <libokang.dev@gmail.com>
Date: Thu, 10 Jun 2021 19:20:58 +0800
Subject: [PATCH 5/5] adapt 1.10 for terway

Signed-off-by: l1b0k <libokang.dev@gmail.com>
---
 pkg/option/config.go | 20 ++++++++++----------
 1 file changed, 10 insertions(+), 10 deletions(-)

diff --git a/pkg/option/config.go b/pkg/option/config.go
index c619a9f599..036987a559 100644
--- a/pkg/option/config.go
+++ b/pkg/option/config.go
@@ -2675,16 +2675,16 @@ func (c *DaemonConfig) Populate() {
 		}
 	}
 
-	switch c.IPAM {
-	case ipamOption.IPAMKubernetes, ipamOption.IPAMClusterPool:
-		if c.EnableIPv4 {
-			c.K8sRequireIPv4PodCIDR = true
-		}
-
-		if c.EnableIPv6 {
-			c.K8sRequireIPv6PodCIDR = true
-		}
-	}
+	//switch c.IPAM {
+	//case ipamOption.IPAMKubernetes, ipamOption.IPAMClusterPool:
+	//	if c.EnableIPv4 {
+	//		c.K8sRequireIPv4PodCIDR = true
+	//	}
+	//
+	//	if c.EnableIPv6 {
+	//		c.K8sRequireIPv6PodCIDR = true
+	//	}
+	//}
 
 	c.KubeProxyReplacementHealthzBindAddr = viper.GetString(KubeProxyReplacementHealthzBindAddr)
 
-- 
2.32.0

