From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: l1b0k <libokang.dev@gmail.com>
Date: Wed, 28 May 2025 22:44:50 +0800
Subject: fix: viper flag

Signed-off-by: l1b0k <libokang.dev@gmail.com>
---
 pkg/option/config.go | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/pkg/option/config.go b/pkg/option/config.go
index ab0731826c..7d8dcd0f8f 100644
--- a/pkg/option/config.go
+++ b/pkg/option/config.go
@@ -3062,7 +3062,7 @@ func (c *DaemonConfig) Populate(vp *viper.Viper) {
 	c.EnableIPIPTermination = vp.GetBool(EnableIPIPTermination)
 	c.EnableUnreachableRoutes = vp.GetBool(EnableUnreachableRoutes)
 	c.EnableNodePort = vp.GetBool(EnableNodePort)
-	c.EnableInClusterLoadBalance = viper.GetBool(EnableInClusterLoadBalance)
+	c.EnableInClusterLoadBalance = vp.GetBool(EnableInClusterLoadBalance)
 	c.EnableSVCSourceRangeCheck = vp.GetBool(EnableSVCSourceRangeCheck)
 	c.EnableHostPort = vp.GetBool(EnableHostPort)
 	c.EnableHostLegacyRouting = vp.GetBool(EnableHostLegacyRouting)
-- 
2.50.1 (Apple Git-155)

