From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: l1b0k <libokang.lbk@alibaba-inc.com>
Date: Thu, 5 Jun 2025 14:16:58 +0800
Subject: fix: src label

Signed-off-by: l1b0k <libokang.dev@gmail.com>
Signed-off-by: l1b0k <libokang.lbk@alibaba-inc.com>
---
 bpf/bpf_lxc.c | 6 ++----
 1 file changed, 2 insertions(+), 4 deletions(-)

diff --git a/bpf/bpf_lxc.c b/bpf/bpf_lxc.c
index dab62a1026..596dcdd30d 100644
--- a/bpf/bpf_lxc.c
+++ b/bpf/bpf_lxc.c
@@ -1793,8 +1793,7 @@ int tail_ipv6_to_endpoint(struct __ctx_buff *ctx)
 				 * as the host. So we can ignore the ipcache
 				 * if it reports the source as HOST_ID.
 				 */
-				if (sec_identity != HOST_ID)
-					src_sec_identity = sec_identity;
+				src_sec_identity = sec_identity;
 			}
 		}
 		cilium_dbg(ctx, info ? DBG_IP_ID_MAP_SUCCEED6 : DBG_IP_ID_MAP_FAILED6,
@@ -2147,8 +2146,7 @@ int tail_ipv4_to_endpoint(struct __ctx_buff *ctx)
 				 * as the host. So we can ignore the ipcache
 				 * if it reports the source as HOST_ID.
 				 */
-				if (sec_identity != HOST_ID)
-					src_sec_identity = sec_identity;
+				src_sec_identity = sec_identity;
 			}
 		}
 		cilium_dbg(ctx, info ? DBG_IP_ID_MAP_SUCCEED4 : DBG_IP_ID_MAP_FAILED4,
-- 
2.50.1 (Apple Git-155)

