From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: l1b0k <libokang.lbk@alibaba-inc.com>
Date: Tue, 21 Nov 2023 14:58:21 +0800
Subject: fix endpoint is lost during restore

Signed-off-by: l1b0k <libokang.lbk@alibaba-inc.com>
---
 pkg/endpoint/restore.go | 22 ++++++++++++++++++----
 1 file changed, 18 insertions(+), 4 deletions(-)

diff --git a/pkg/endpoint/restore.go b/pkg/endpoint/restore.go
index 21124db6fb..b02803db11 100644
--- a/pkg/endpoint/restore.go
+++ b/pkg/endpoint/restore.go
@@ -113,8 +113,24 @@ func ReadEPsFromDirNames(ctx context.Context, owner regeneration.Owner, policyGe
 		}
 
 		if err := headerFileExists(); err != nil {
-			scopedLog.WithError(err).Warn("C header file not found. Ignoring endpoint")
-			continue
+			// synchronizeDirectories() logic would rename origDir (xxx) to backupDir (xxx_stale)
+			// is is possible previous is a not complete restore directory
+			// try restore if backup exists
+			backupDir := epDir + backupDirectorySuffix
+			origDir := epDir
+			_, err = os.Stat(backupDir)
+			if err != nil {
+				scopedLog.WithError(err).Warn("C header file not found. Ignoring endpoint")
+				continue
+			}
+			log.Warnf("unable to locate C header file %s. Trying to restore from backup", cHeaderFile)
+
+			_ = os.Mkdir(origDir, os.ModeDir)
+			err = moveNewFilesTo(backupDir, origDir)
+			if err != nil {
+				log.WithError(err).Warnf("unable to copy old bpf object "+
+					"files from %s into the new directory %s.", backupDir, origDir)
+			}
 		}
 
 		// This symlink is only needed when upgrading from a pre-1.11 Cilium
@@ -407,14 +423,12 @@ func (e *Endpoint) toSerializedEndpoint() *serializableEndpoint {
 // serializableEndpoint contains the fields from an Endpoint which are needed to be
 // restored if cilium-agent restarts.
 //
-//
 // WARNING - STABLE API
 // This structure is written as JSON to StateDir/{ID}/ep_config.h to allow to
 // restore endpoints when the agent is being restarted. The restore operation
 // will read the file and re-create all endpoints with all fields which are not
 // marked as private to JSON marshal. Do NOT modify this structure in ways which
 // is not JSON forward compatible.
-//
 type serializableEndpoint struct {
 	// ID of the endpoint, unique in the scope of the node
 	ID uint16
-- 
2.39.5 (Apple Git-154)

