Index: daemon/cmd/endpoint.go
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/daemon/cmd/endpoint.go b/daemon/cmd/endpoint.go
--- a/daemon/cmd/endpoint.go	(revision 952d9d33740fc77e1a94d5a52b38d2a313e5c570)
+++ b/daemon/cmd/endpoint.go	(revision d25122164324795e85c9dc59a175b903779cbd82)
@@ -361,7 +361,9 @@
 		if err != nil {
 			return invalidDataError(ep, err)
 		} else if oldEp != nil {
-			return invalidDataError(ep, fmt.Errorf("IP %s is already in use", id))
+			oldEp.Logger("api").Warning("endpoint conflicted, deleting")
+			d.deleteEndpoint(oldEp)
+			// return invalidDataError(ep, fmt.Errorf("IP %s is already in use", id))
 		}
 	}
 
